// @react-three/drei/native/index.js
import 'blob-polyfill';
import '../helpers/base64.js';
import '../assets/cloud.base64.js';
import '@babel/runtime/helpers/esm/extends';
import '@babel/runtime/helpers/esm/objectWithoutPropertiesLoose';
import 'react';
import 'react-three-fiber';
export { B as Billboard, W as Box, o as Center, X as Circle, C as Cloud, Y as Cone, V as ContactShadows, G as CurveModifier, Z as Cylinder, D as Detailed, c as DeviceOrientationControls, a8 as Dodecahedron, a as Effects, E as Environment, a9 as Extrude, F as FlyControls, a6 as Icosahedron, aa as Lathe, L as Line, M as MapControls, H as MeshDistortMaterial, I as MeshWobbleMaterial, a7 as Octahedron, d as OrbitControls, O as OrthographicCamera, ab as Parametric, b as PerspectiveCamera, $ as Plane, g as PointerLockControls, a5 as Polyhedron, P as PositionalAudio, p as Preload, R as Reflector, a4 as Ring, ac as RoundedBox, ad as ScreenQuad, S as Shadow, N as Sky, _ as Sphere, U as Stars, n as Stats, a3 as Tetrahedron, T as Text, a1 as Torus, a2 as TorusKnot, e as TrackballControls, f as TransformControls, a0 as Tube, K as calcPosFromAngles, i as isWebGL2Available, m as meshBounds, J as shaderMaterial, Q as softShadows, w as useAnimations, q as useAspect, r as useCamera, v as useContextBridge, u as useCubeTexture, s as useDetectGPU, A as useEdgeSplit, x as useFBO, h as useFBX, j as useGLTF, t as useHelper, ae as useMatcapTexture, af as useNormalTexture, k as useProgress, y as useSimplification, z as useTessellation, l as useTexture } from '../useNormalTexture-39fc8195.js';
import 'react-merge-refs';
import 'three';
import 'three/examples/jsm/lines/LineGeometry';
import 'three/examples/jsm/lines/LineMaterial';
import 'three/examples/jsm/lines/Line2';
import 'troika-three-text';
import 'three/examples/jsm/loaders/RGBELoader';
import 'use-asset';
import '../helpers/environment-assets.js';
import 'three/examples/jsm/postprocessing/EffectComposer';
import 'three/examples/jsm/postprocessing/RenderPass';
import 'three/examples/jsm/postprocessing/ShaderPass';
import 'three/examples/jsm/shaders/GammaCorrectionShader';
import 'three/examples/jsm/controls/DeviceOrientationControls';
import 'three/examples/jsm/controls/FlyControls';
import 'three/examples/jsm/controls/OrbitControls';
import 'three/examples/jsm/controls/TrackballControls';
import 'three/examples/jsm/controls/TransformControls';
import 'lodash.pick';
import 'lodash.omit';
import 'three/examples/jsm/controls/PointerLockControls';
import 'three/examples/jsm/loaders/FBXLoader';
import 'three/examples/jsm/loaders/GLTFLoader';
import 'three/examples/jsm/loaders/DRACOLoader';
import 'three/examples/jsm/libs/meshopt_decoder.module.js';
import 'zustand';
import '@babel/runtime/helpers/esm/defineProperty';
export { B as BlurPass, M as MeshReflectorMaterial } from '../MeshReflectorMaterial-f302adf4.js';
import 'three/examples/js/libs/stats.min';
import 'detect-gpu';
import 'three/examples/jsm/modifiers/SimplifyModifier';
import 'three/examples/jsm/modifiers/TessellateModifier';
import 'three/examples/jsm/modifiers/EdgeSplitModifier';
import 'three/examples/jsm/modifiers/CurveModifier';
import 'three/examples/jsm/objects/Sky';
import 'three/examples/jsm/shaders/HorizontalBlurShader';
import 'three/examples/jsm/shaders/VerticalBlurShader';
import '../helpers/matcap-assets.js';
import '../helpers/normal-assets.js';
